libfetchers: Always include (empty) submodule directories with `git+file` schema

juliankuners opened the pull request:
## Motivation

With `submodules=0`, the `git+file` fetcher diverges in behavior compared to `github` and `git+https` fetchers.

In particular, `git+https` and `github` fetchers always include empty submodule directories, while `git+file` currently does not. This can cause diverging builds, which can lead to binary cache misses in turn. In addition, the fetcher cache can be populated with diverging entries causing assertion errors in `copyInputToStore`, see related but duplicated issue #14317.

This PR changes the `git+https` to always include the submodule directories, instead of only when `submodules=1`. This is done by allowing accessors access to submodule directory paths, but not their children nodes.

Also, it is assumed that local checkouts always contain the respective empty submodule directory, otherwise the local checkout would be in a corrupted state, in which case the directory would also be omitted from the derivation output.

## Context

Closes: #13698

---

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

juliankuners committed:
always include (empty) submodule directories with git+file schema

With `submodules=0`, the `git+file` fetcher diverges in behaviour
compared to `github` and `git+https` fetchers.

In particular, `git+https` and `github` fetchers always include
empty submodule directories, while `git+file` does not. This can
cause diverging builds due to different source files. In addition,
the fetcher cache can be populated with diverging entries causing
assertion errors in `copyInputToStore`.

juliankuners committed:
Merge branch 'master' into git-file-submodule-dirs

juliankuners committed:
Merge branch 'master' into git-file-submodule-dirs

juliankuners committed:
Merge branch 'master' into git-file-submodule-dirs

xokdvium commented:
Overall this seems quite safe, because local repos should not end up in lock files where hash mismatches would be a concern. Bringing this in-line with other git fetchers is the right call imo

xokdvium commented on src/libfetchers/include/nix/fetchers/git-utils.hh:
```diff
@@ -82,6 +84,9 @@ struct GitRepo
         std::vector<Submodule> submodules;
+
+        /* The empty directories that must be present in the nix derivation output. */
+        boost::unordered_flat_set<CanonPath> emptyDirs;
```
The comment is slightly misleading. There are no derivations to speak of here.

juliankuners committed:
add regression test for empty git submodule dirs

The associated PR https://github.com/NixOS/nix/pull/14779 adjusted the
behaviour of the `git+file` fetcher to match the `git+https`
and `github` fetchers when copying empty submodule directories
that are generated by git,
see issue https://github.com/NixOS/nix/issues/13698

This regression test tests that empty git submodule directories are
indeed copied into the nix store when using the `git+file` fetcher.

juliankuners committed:
adjust comment for attribute `GitRepo::WorkdirInfo::emptyDirs`

juliankuners committed:
Merge branch 'master' into git-file-submodule-dirs

juliankuners commented:
> It is essential that this has accompanying tests. Probably we should also compare fetching with `_NIX_FORCE_HTTP=1` and without it too.

I added a regression test to `tests/functional/fetchGit.sh`. This test checks whether the `git+file` fetcher not correctly includes an empty submodule directory. It tested that the test properly fails when leaving out my fix commit.

In this test I do not make use of `_NIX_FORCE_HTTP=1`, since I noticed that using `_NIX_FORCE_HTTP=1` causes the resulting nix path to be completely empty, with both this PR applied and not applied. I suspect this is due to my use of a local path submodule, where `url` in `.gitmodules` is a path instead of a git URL. Specifically, this git repository has a different path than specified by `path` in `.gitmodules` and is located outside the respective repository. I've done this since otherwise the regression test actually does not fail since the directory is checked in as part of the parent repository. Using such a submodule requires git to be configured with `protocol.file.allow=always`.

juliankuners committed:
add `_NIX_FORCE_HTTP` variant to new test for empty submodule dirs

juliankuners commented:
Never mind, I was missing `--tarball-ttl 0` and therefore observed a cached empty store output. Using `_NIX_FORCE_HTTP=1` works just fine and I added a respective variant to the test.

I now test with both `_NIX_FORCE_HTTP=1` and without whether an empty submodule directory is copied into a store path and whether that directory is empty as `submodules=0` in this test. I could readjust this to simply compare both outputs, but I went for being more explicit with what the test tests for specifically. Let me know what you think.

@xokdvium Should be good for re-review now. Thanks for taking a look at this PR!

juliankuners committed:
Merge branch 'master' into git-file-submodule-dirs

xokdvium commented on src/libfetchers/git-utils.cc:
```diff
@@ -1393,11 +1393,16 @@ ref<SourceAccessor> GitRepoImpl::getAccessor(
                                            std::set<CanonPath>{wd.files},
-                                           // Always allow access to the root, but not its children.
-                                           boost::unordered_flat_set<CanonPath>{CanonPath::root},
+                                           std::move(allowedPrefixes),
```
Do I understand correctly that subdirectories would be excluded when the working directory has submodule checkouts because `wd.files` wouldn't list those files?

juliankuners replied:
Hmm, I seem to have messed up the naming of this variable. I pushed a commit that renames the variable `allowedPrefixes` to `allowedPaths`, which now properly reflects the variable names used in `AllowListSourceAccessor::create`.

Yes, `wd.files` does not list the submodule directories and therefore the directories would not be copied over. However, if `wd.files` were to list submodule directories, their content would also be copied over, which is not the intended behaviour when `submodules=0`. Hence the differentiation here.

juliankuners committed:
rename a variable to `allowedPaths`

juliankuners committed:
Merge remote-tracking branch 'origin/master' into juliankuners-git-file-submodule-dirs

xokdvium commented on src/libfetchers/git-utils.cc:
Having thought about it a bit more, those directories aren't strictly guaranteed to be empty though? We might want to enforce this by hiding the actual contents to be empty directories (something like memory source accessor with empty root directory?) at those locations?

juliankuners replied:
The PR makes use of `allowedPaths` instead of `allowedPrefixes`. This should, if I understand correclty, ensure that only the directory itself is accessible and not it's children, [see here](https://github.com/juliankuners/nix/blob/1c9d963c80ae59ac0adaf4da3106ee47b5fbbfaf/src/libfetchers/git-utils.cc#L1454).

E.g., `submodules=1` instead makes use of `allowedPrefixes` and additionally adds mounts for the submodule directories, [see here](https://github.com/juliankuners/nix/blob/1c9d963c80ae59ac0adaf4da3106ee47b5fbbfaf/src/libfetchers/git.cc#L988).

xokdvium replied:
Ah right. Though there's a separate concern if they are listed in the modules list, but don't have a corresponding commit object (though that would be more of a separate fix along the lines of https://github.com/NixOS/nix/issues/15423).

juliankuners committed:
change boost `unordered_flat_set` to `std`

juliankuners committed:
Merge branch 'master' into git-file-submodule-dirs
